File: leet_BTREE_2196_create_btree.cc
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    TreeNode* createBinaryTree(vector<vector<int>>& descriptions)
    {
        // description -> parent-child relation
        unordered_map<int, TreeNode *> nodes;
        unordered_set<int> ischild;
        for (auto t : descriptions)
        {
            int parent = t[0];
            int child = t[1];
            int isleft = t[2];
            if (nodes[parent] == nullptr)
                nodes[parent] = new TreeNode(parent);
            if (nodes[child] == nullptr)
                nodes[child] = new TreeNode(child);
            if (isleft)
                nodes[parent]->left = nodes[child];
            else
                nodes[parent]->right = nodes[child];
            ischild.insert(child);
        }
        TreeNode * res = nullptr;
        for (auto p : nodes)
        {
            if (ischild.find(p.first) == end(ischild))
            {
                res = p.second;
                break ;
            };
        }
        return res;
    }
};

File: leet_BTREE_2196_create_btree.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func createBinaryTree(descriptions [][]int) *TreeNode {

    nodes := make(map[int] *TreeNode)
    ischild := make(map[int]bool)
    for _, t := range descriptions {
        parent := t[0]
        child := t[1]
        isleft := t[2]
        if _, isset := nodes[parent]; ! isset {
            nodes[parent] = & TreeNode{ Val : parent }
        }
        if _, isset := nodes[child]; ! isset {
            nodes[child] = & TreeNode{ Val : child }
        }
        if isleft == 1 {
            nodes[parent].Left = nodes[child]
        } else {
            nodes[parent].Right = nodes[child]
        }
        ischild[child] = true
    }
    var res * TreeNode
    for node, realnode := range nodes {
        if ! ischild[node] {
            res = realnode
            break
        }
    }
    return res

}
